File: packages/guilded/src/outgoing.ts
import type { Client } from '@jersey/guildapi';
import type { ChatEmbed } from '@jersey/guilded-api-types';
import type { message } from '@jersey/lightning';
import { fetch_author } from './incoming.ts';

type guilded_payload = {
	content?: string;
	embeds?: ChatEmbed[];
	replyMessageIds?: string[];
	avatar_url?: string;
	username?: string;
};

const username = /^[a-zA-Z0-9_ ()-]{1,25}$/ms;

function get_name(msg: message): string {
	if (username.test(msg.author.username)) {
		return msg.author.username;
	} else if (username.test(msg.author.rawname)) {
		return msg.author.rawname;
	} else {
		return `${msg.author.id}`;
	}
}

async function fetch_reply(
	msg: message,
	client: Client,
): Promise<ChatEmbed | undefined> {
	if (!msg.reply_id) return;

	try {
		const replied_to = await client.request(
			'get',
			`/channels/${msg.channel_id}/messages/${msg.reply_id}`,
			undefined,
		);

		const author = await fetch_author(replied_to.message, client);

		return {
			author: {
				name: `reply to ${author.username}`,
				icon_url: author.profile,
			},
			description: replied_to.message.content,
		};
	} catch {
		return;
	}
}

export async function get_outgoing(
	msg: message,
	client: Client,
	limitMentions?: boolean,
): Promise<guilded_payload> {
	const message: guilded_payload = {
		content: msg.content,
		avatar_url: msg.author.profile,
		username: get_name(msg),
		embeds: msg.embeds?.map((i) => {
			return {
				...i,
				fields: i.fields
					? i.fields.map((j) => {
						return {
							...j,
							inline: j.inline ?? false,
						};
					})
					: undefined,
				timestamp: i.timestamp ? i.timestamp.toString() : undefined,
			};
		}),
	};

	if (msg.reply_id) {
		const embed = await fetch_reply(msg, client);

		if (embed) {
			if (!message.embeds) message.embeds = [];
			message.embeds.push(embed);
		}
	}

	if (msg.attachments?.length) {
		if (!message.embeds) message.embeds = [];
		message.embeds.push({
			title: 'attachments',
			description: msg.attachments
				.slice(0, 5)
				.map((a) => {
					return `![${a.alt || a.name}](${a.file})`;
				})
				.join('\n'),
		});
	}

	if (!message.content && !message.embeds) message.content = '\u2800';

	if (limitMentions && message.content) {
		message.content = message.content.replace(/@everyone/gi, '(a)everyone');
		message.content = message.content.replace(/@here/gi, '(a)here');
	}

	return message;
}

File: pkg/platforms/guilded/outgoing.go
package guilded

import (
	"encoding/json"
	"log"
	"regexp"
	"strings"

	"github.com/williamhorning/lightning/pkg/lightning"
)

var usernameRegex = regexp.MustCompile(`(?ms)^[a-zA-Z0-9_ ()-]{1,25}$`)

func (p *guildedPlugin) getOutgoingMessage(message *lightning.Message, opts *lightning.SendOptions) *guildedPayload {
	base := &guildedPayload{
		Content:         message.Content,
		ReplyMessageIDs: message.RepliedTo,
		Embeds:          p.appendReplyEmbed(message),
	}

	if message.Author != nil {
		if message.Author.ProfilePicture != "" {
			base.AvatarURL = message.Author.ProfilePicture
		}

		base.Username = message.Author.ID

		if usernameRegex.MatchString(message.Author.Username) {
			base.Username = message.Author.Username
		}

		if usernameRegex.MatchString(message.Author.Nickname) {
			base.Username = message.Author.Nickname
		}
	}

	if len(base.Content) == 0 && len(base.Embeds) == 0 {
		base.Content = "\u2800"
	}

	if opts != nil && !opts.AllowEveryonePings {
		base.Content = strings.ReplaceAll(base.Content, "@everyone", "@\u2800everyone")
		base.Content = strings.ReplaceAll(base.Content, "@here", "@\u2800here")
	}

	return base
}

func (p *guildedPlugin) appendReplyEmbed(message *lightning.Message) []lightning.Embed {
	resp, err := guildedMakeRequest(p.token, "GET",
		"/channels/"+message.ChannelID+"/messages/"+message.RepliedTo[0], nil)
	if err != nil {
		return message.Embeds
	}

	if resp.Body.Close() != nil {
		log.Println("guilded: failed to close request body when getting reply embed")
	}

	var messageResp guildedChatMessageWrapper
	if json.NewDecoder(resp.Body).Decode(&messageResp) != nil {
		return message.Embeds
	}

	author := p.getIncomingAuthor(&messageResp.Message)
	title := "reply to " + author.Nickname

	return append(message.Embeds, lightning.Embed{
		Author: &lightning.EmbedAuthor{
			Name:    title,
			IconURL: author.ProfilePicture,
		},
		Description: messageResp.Message.Content,
	})
}
